sdk/core/Azure.Core.TestFramework/src/TestEnvironment.cs

```diff
@@ -201,7 +202,24 @@ public async ValueTask WaitForEnvironmentAsync()
         {
             if (GlobalIsRunningInCI && Mode == RecordedTestMode.Live)
             {
-                await s_environmentStateCache.GetOrAdd(GetType(), t => WaitForEnvironmentInternalAsync());
+                // GetOrAdd from ConcurrentDictionary isn't atomic, hence double-checked-lock.
+                if (!s_environmentStateCache.TryGetValue(GetType(), out Task task))
```

Comment:
Do we need a semaphore? Can we just lock?

Reply:
good point, we don't await in critical section, should be fine.

Reply:
We need to remove this one and always access the dictionary under the lock.

Reply:
Shouldn't be too bad. As everyone would have to wait on the task anyway.